lib/6to5/traverse/index.js

```diff
@@ -50,104 +57,139 @@ TraversalContext.prototype.maybeReplace = function (result, obj, key, node) {
+
+  if (result) {
+    flatten = replaceNode(obj, key, node, result);
+    node = result;
+
+    if (flatten) {
+      this.didFlatten = true;
+    }
+  }
+
+  if (this.didRemove) {
+    obj[key] = null;
+    this.didFlatten = true;
+  }
+
+  return node;
+};
+
+TraversalContext.prototype.exitNode = function (obj, key, node, exit, parent, scope, state) {
+  var result = exit(node, parent, scope, this, state);
+  var flatten = false;
+
+  if (result) {
+    flatten = replaceNode(obj, key, node, result);
+    node = result;
+
+    if (flatten) {
+      this.didFlatten = true;
+    }
   }
 
   return node;
 };
 
-TraversalContext.prototype.visit = function (obj, key, opts, scope, parent) {
+TraversalContext.prototype.visitNode = function (obj, key, opts, scope, parent, state) {
+  this.reset();
+
   var node = obj[key];
-  if (!node) return;
 
   // type is blacklisted
-  if (opts.blacklist && opts.blacklist.indexOf(node.type) > -1) return;
+  if (opts.blacklist && opts.blacklist.indexOf(node.type) > -1)
+    return;
 
-  var result;
   var ourScope = scope;
-  if (t.isScope(node)) ourScope = new Scope(node, scope);
+  if (t.isScope(node))
+    ourScope = new Scope(node, scope);
 
-  // enter
-  if (opts.enter) {
-    result = opts.enter.call(this, node, parent, ourScope);
-    node = this.maybeReplace(result, obj, key, node);
+  node = this.enterNode(obj, key, node, opts.enter, parent, ourScope, state);
 
-    if (this.didRemove) {
-      obj[key] = null;
-      this.flatten();
-    }
+  if (this.didSkip)
+    return this.didStop;
+
+  traverseNode(node, opts, ourScope, state);
+  this.exitNode(obj, key, node, opts.exit, parent, ourScope, state);
 
-    // stop traversal
-    if (this.didSkip) return;
+  return this.didStop;
+};
+
+TraversalContext.prototype.visit = function (node, key, opts, scope, state) {
+  var nodes = node[key];
+  if (!nodes) return;
+
+  if (!Array.isArray(nodes)) {
+    return this.visitNode(node, key, opts, scope, node, state);
   }
 
-  // traverse node
-  traverse(node, opts, ourScope);
+  if (nodes.length === 0) {
+    return;
+  }
 
-  // exit
-  if (opts.exit) {
-    result = opts.exit.call(this, node, parent, ourScope);
-    node = this.maybeReplace(result, obj, key, node);
+  for (var k = 0; k < nodes.length; k++) {
+    if (nodes[k] && this.visitNode(nodes, k, opts, scope, node, state))
+      return true;
   }
-};
 
-function traverse(parent, opts, scope) {
-  // falsy node
-  if (!parent) return;
+  if (this.didFlatten) {
+    node[key] = _.flatten(node[key]);
 
-  // array of nodes
-  if (Array.isArray(parent)) {
-    for (var i = 0; i < parent.length; i++)
-      traverse(parent[i], opts, scope);
-    return;
+    if (key === "body") {
+      // we can safely compact this
+      node[key] = _.compact(node[key]);
+    }
   }
+};
 
-  // unknown node type to traverse
-  var keys = t.VISITOR_KEYS[parent.type];
+function traverseNode(node, opts, scope, state) {
+  var keys = t.VISITOR_KEYS[node.type];
```

Comment:
an additional thing that we can do to VISITOR_KEYS is ensure it is of a `null` prototype, this way property misses don't need to walk the prototype chain. I _believe_ the last time i investigated this, the fastProperty lookup table didn't handle misses.